In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('Fraud.csv')

In [3]:
df.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0             0.0        0               0  
1  M2044282225             0.0             0.0        0               0  
2   C553264065             0.0             0.0        1               0  
3    C38997010         21182.0             0.0        1               0  
4  M1230701703             0.0             0.0        0               0

In [4]:
df.tail()

step      type      amount     nameOrig  oldbalanceOrg  \
6362615   743  CASH_OUT   339682.13   C786484425      339682.13   
6362616   743  TRANSFER  6311409.28  C1529008245     6311409.28   
6362617   743  CASH_OUT  6311409.28  C1162922333     6311409.28   
6362618   743  TRANSFER   850002.52  C1685995037      850002.52   
6362619   743  CASH_OUT   850002.52  C1280323807      850002.52   

         newbalanceOrig     nameDest  oldbalanceDest  newbalanceDest  isFraud  \
6362615             0.0   C776919290            0.00       339682.13        1   
6362616             0.0  C1881841831            0.00            0.00        1   
6362617             0.0  C1365125890        68488.84      6379898.11        1   
6362618             0.0  C2080388513            0.00            0.00        1   
6362619             0.0   C873221189      6510099.11      7360101.63        1   

         isFlaggedFraud  
6362615               0  
6362616               0  
6362617               0  
6362618               0  
6362619               0

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


In [6]:
df.shape

(6362620, 11)

# Analysis Data And EDA Processing

* cleaning the data
* findind missing values
* finding outliers
* finding multi-collinearity

In [7]:
df.isnull().sum()

step              0
type              0
amount            0
nameOrig          0
oldbalanceOrg     0
newbalanceOrig    0
nameDest          0
oldbalanceDest    0
newbalanceDest    0
isFraud           0
isFlaggedFraud    0
dtype: int64

No null values in the data set to be fund

In [8]:
df.duplicated().sum()

0

therfore no duplicate rows are present in it

In [9]:
df.isFraud.value_counts()

isFraud
0    6354407
1       8213
Name: count, dtype: int64

* this tells us that there are 6354407 no fraud value
* and 8213 are frauded values

In [10]:
df.isFlaggedFraud.value_counts()

isFlaggedFraud
0    6362604
1         16
Name: count, dtype: int64

* isFlaggedFraud tells or indicates a suspicion or potential fraud case, but it has not been confirmed therefore 16 case canbe frauded but not conferamtion

# Data Preprocessing

In [11]:
df.columns

Index(['step', 'type', 'amount', 'nameOrig', 'oldbalanceOrg', 'newbalanceOrig',
       'nameDest', 'oldbalanceDest', 'newbalanceDest', 'isFraud',
       'isFlaggedFraud'],
      dtype='object')

as we can see 'nameOrig' and 'nameDest' is not use for us so we drop these columns, also this help to optimize operations for futher analysis

In [12]:
df = df.drop(['nameOrig','nameDest'],axis = 1)

In [13]:
df.columns

Index(['step', 'type', 'amount', 'oldbalanceOrg', 'newbalanceOrig',
       'oldbalanceDest', 'newbalanceDest', 'isFraud', 'isFlaggedFraud'],
      dtype='object')

### Now we do some label encoding
* this is also called hot label encoding
* it is a algorithm which work on the binary of the system of the data
* it deals with 0 and 1

In [14]:
from sklearn import preprocessing

label_encoder = preprocessing.LabelEncoder()
df['type'] = label_encoder.fit_transform(df['type'])

* In above cell we import `LabelEncoder` from sklearn preprocessing module
* this help to convert categorical data in the type cloumne of datafram df into numerical data
* this is necessary for machine learning models, which is generally require numerical input
* it help in `machine learning compatibility`, `optimize performance`, `data preprocessing`


In [15]:
df.head(10)

step  type    amount  oldbalanceOrg  newbalanceOrig  oldbalanceDest  \
0     1     3   9839.64      170136.00       160296.36             0.0   
1     1     3   1864.28       21249.00        19384.72             0.0   
2     1     4    181.00         181.00            0.00             0.0   
3     1     1    181.00         181.00            0.00         21182.0   
4     1     3  11668.14       41554.00        29885.86             0.0   
5     1     3   7817.71       53860.00        46042.29             0.0   
6     1     3   7107.77      183195.00       176087.23             0.0   
7     1     3   7861.64      176087.23       168225.59             0.0   
8     1     3   4024.36        2671.00            0.00             0.0   
9     1     2   5337.77       41720.00        36382.23         41898.0   

   newbalanceDest  isFraud  isFlaggedFraud  
0            0.00        0               0  
1            0.00        0               0  
2            0.00        1               0  
3            0.00        1               0  
4            0.00        0               0  
5            0.00        0               0  
6            0.00        0               0  
7            0.00        0               0  
8            0.00        0               0  
9        40348.79        0               0

# Training and Testing

In [16]:
X, y = df.loc[:,df.columns != 'isFraud'],df['isFraud']

Feature Selection, Training Models

In [17]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.40, random_state=42)

Model Training, Model Testing, Avoid Overfitting

#### Standardization

In [18]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

Improves Model Performance, Faster Convergence, Equal Contribution

In [24]:
from sklearn.naive_bayes import GaussianNB
from sklearn import metrics

gnb = GaussianNB()

gnb.fit(X_train, y_train)

y_pred = gnb.predict(X_test)


print("Accuracy: ", metrics.accuracy_score(y_test, y_pred))

Accuracy:  0.9955882953877491


In [25]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)
y_pred = classifier.predict(X_test)

print("Accuracy : ", metrics.accuracy_score(y_test, y_pred))

Accuracy :  0.9991882274911907
